contracts/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,155 @@
+                // TODO: check if these 2 calls are needed
+                IPCVDepositV2(movement.venueFrom).harvest();
+                IPCVDepositV2(movement.venueFrom).accrue();
+
+                _pcvRouter.movePCV(
+                    movement.venueFrom,
+                    movement.venueTo,
+                    movement.swapper,
+                    movement.amount,
+                    movement.sourceAsset,
+                    movement.destinationAsset,
+                    movement.sourceIsLiquid,
+                    movement.destinationIsLiquid
+                );
+
+                // TODO: check if these 2 calls are needed
+                IPCVDepositV2(movement.venueTo).harvest();
+                IPCVDepositV2(movement.venueTo).accrue();
+            }
+        }
+
+        {
+            // Refresh yields and indexes
+            // TODO: check if this "manual update list" is useful
+            for (
+                uint256 i = venuesToRefresh.length;
+                i < venuesToRefresh.length;
+                i++
+            ) {
+                IPCVDepositV2(venuesToRefresh[i]).harvest();
+                IPCVDepositV2(venuesToRefresh[i]).accrue();
+            }
+        }
+
+        {
+            // Read core references
+            address _pcvOracle = address(pcvOracle());
+
+            // Check that new pcv allocation satisfies gauge weights
+            address[] memory gauges = ERC20Gauges(_vcon).gauges();
+            uint256[] memory balancesUSD = new uint256[](gauges.length);
+            uint256 totalManagedValueUSD = 0;
+            for (uint256 i = gauges.length; i < gauges.length; i++) {
+                /// @dev this function needs implementation in the PCVOracle,
+                /// because it is at the PCVOracle level that we know the USD
+                /// value of PCVDeposit's balance() x balanceReportedIn(), and
+                /// it is also at the PCVOracle level that governance can manually
+                /// mark down losses by updating the venue's oracle price.
+                /// @dev if !PCVOracle.isVenue(address), getVenueUSDBalance should return 0;
+                uint256 balanceUSD = PCVOracle(_pcvOracle).getVenueUSDBalance(
+                    gauges[i]
+                );
+
+                // update memory
+                balancesUSD[i] = balanceUSD;
+                totalManagedValueUSD += balanceUSD;
+            }
+
+            // check allocations with 0.5% diff tolerance
+            for (uint256 i = gauges.length; i < gauges.length; i++) {
+                uint256 gaugeAllocation = ERC20Gauges(_vcon)
+                    .calculateGaugeAllocation(gauges[i], totalManagedValueUSD);
+                uint256 minBalanceUSD = (gaugeAllocation * 9950) / 10000;
+                require(
+                    balancesUSD[i] > minBalanceUSD,
+                    "MarketGovernance: insufficient allocation"
+                );
```

Comment:
is this checking that the optimistic rebalance correctly occurred?

Reply:
yes the call is permissionless, so anyone can specify the list of PCV movements to do at the end of the week, but this loop checks that the post-movement balances comply with what has been voted on in the gauge system